tuxemon/states/combat/combat.py

```diff
@@ -1193,19 +1194,53 @@ def trigger_player_run(self, player: NPC) -> None:
         self.players.remove(player)
 
     def evolve(self) -> None:
+        self.client.pop_state()
         for monster in self.players[0].monsters:
             for evolution in monster.evolutions:
-                # check the path field, path field signals evolution item based
-                if not evolution.path:
-                    if evolution.at_level <= monster.level:
-                        logger.info(
-                            "{} evolved into {}!".format(
-                                monster.name, evolution.monster_slug
+                if evolution.at_level <= monster.level:
+                    tools.open_dialog(
+                        local_session,
+                        [
+                            T.format(
+                                "evolution_confirmation",
+                                {"name": monster.name},
                             )
+                        ],
+                    )
+                    tools.open_choice_dialog(
+                        local_session,
+                        menu=(
+                            (
+                                "yes",
+                                T.translate("yes"),
+                                partial(
+                                    self.positive_answer, monster.instance_id
+                                ),
+                            ),
+                            ("no", T.translate("no"), self.negative_answer),
+                        ),
+                    )
+
+    def positive_answer(self, instance_id: uuid.UUID) -> None:
+        self.client.pop_state()
+        self.client.pop_state()
+        monster = self.players[0].find_monster_by_id(instance_id)
```

Comment:
On the right lines, but you could have just passed the monster in.

Reply:
No problem